src/crimson/os/seastore/segment_cleaner.cc

```diff
@@ -435,30 +488,103 @@ SegmentCleaner::init_segments_ret SegmentCleaner::init_segments() {
 	auto segment_id = it.first;
 	return scanner->read_segment_header(
 	  segment_id
-	).safe_then([&segment_set, segment_id, this](auto header) {
-	  if (header.out_of_line) {
-	    logger().debug(
-	      "ExtentReader::init_segments: out-of-line segment {}",
-	      segment_id);
-	    init_mark_segment_closed(
-	      segment_id,
-	      header.journal_segment_seq,
-	      true);
-	  } else {
-	    logger().debug(
-	      "ExtentReader::init_segments: journal segment {}",
-	      segment_id);
-	    segment_set.emplace_back(std::make_pair(segment_id, std::move(header)));
-	  }
-	  return seastar::now();
+	).safe_then([&segment_set, segment_id, this](auto header)
+	  -> ExtentReader::scan_extents_ertr::future<> {
+	  return scanner->read_segment_tail(
+	    segment_id
+	  ).safe_then([this, segment_id, &segment_set, header](auto tail)
+	    -> ExtentReader::read_segment_tail_ertr::future<> {
+	    if (tail.segment_nonce != header.segment_nonce) {
+	      return crimson::ct_error::enodata::make();
+	    }
+	    seastar::lowres_system_clock::time_point last_modified(
+	      seastar::lowres_system_clock::duration(tail.last_modified));
+	    seastar::lowres_system_clock::time_point last_rewritten(
+	      seastar::lowres_system_clock::duration(tail.last_rewritten));
+	    if (segments[segment_id].last_modified < last_modified) {
+	      segments[segment_id].last_modified = last_modified;
+	    }
+	    if (segments[segment_id].last_rewritten < last_rewritten) {
+	      segments[segment_id].last_rewritten = last_rewritten;
+	    }
+	    if (header.out_of_line) {
+	      logger().info(
+		"ExtentReader::init_segments: out-of-line segment {}",
+		segment_id);
+	      init_mark_segment_closed(
+		segment_id,
+		header.journal_segment_seq,
+		true);
+	    } else {
+	      logger().info(
+		"ExtentReader::init_segments: journal segment {}",
+		segment_id);
+	      segment_set.emplace_back(std::make_pair(segment_id, std::move(header)));
+	    }
+	    return seastar::now();
+	  }).handle_error(
+	    crimson::ct_error::enodata::handle(
+	      [this, header, segment_id, &segment_set](auto)
+	      -> ExtentReader::scan_extents_ertr::future<> {
+	      if (header.out_of_line) {
+		logger().info(
+		  "ExtentReader::init_segments: out-of-line segment {}",
+		  segment_id);
+		return seastar::do_with(
+		  scan_valid_records_cursor({
+		    segments[segment_id].journal_segment_seq,
+		    paddr_t::make_seg_paddr(segment_id, 0)}),
+		  [this, segment_id](auto& cursor) {
+		  return scanner->scan_extents(
+		    cursor,
+		    segments[segment_id.device_id()]->segment_size
```

Comment:
Add a helper to extent_reader.h that simply reads the entire specified segment, that would avoid needing the scan_valid_records_cursor setup here.